guidance/_parser.py

```diff
@@ -133,6 +97,54 @@ def _parse(
         return response
 
 
+def process_prompt(prompt_tokens: Sequence[int], ll_interpreter: llguidance.LLInterpreter, bos_token_id: Optional[int]=None) -> list[int]:
+    # Allows ll_interpreter to make adjustments to prompt tokens, such as token healing
+    processed_tokens = ll_interpreter.process_prompt(prompt_tokens)
+    if (
+        bos_token_id is not None
+        and prompt_tokens[:1] != [bos_token_id]
+    ):
+        # add the beginning of sequence token if needed
+        processed_tokens = [bos_token_id] + processed_tokens
+
+    return processed_tokens
+
+
+def process_grammar(grammar: Union[GrammarFunction, str]) -> str:
+    if isinstance(grammar, GrammarFunction):
+        # we can't have a terminal as the root
+        if isinstance(grammar, Terminal):
+            grammar = Join([grammar])
+        return json.dumps(grammar.ll_serialize())
+    else:
+        return grammar
+
+
+def create_token_parser(
+    grammar: Union[GrammarFunction, str],
+    tokenizer: Tokenizer,
+    prompt: bytes = b"",
+    ensure_bos_token: bool = True,
+    trace: bool = False
+) -> TokenParser:
+    serialized_grammar = process_grammar(grammar)
+    ll_tokenizer = llguidance.LLTokenizer(
+        llguidance.TokenizerWrapper(tokenizer)
+    )
+    ll_interpreter = llguidance.LLInterpreter(
+        ll_tokenizer,
+        serialized_grammar,
+        log_level=2 if trace else int(os.environ.get("LLGUIDANCE_LOG_LEVEL", "1")),
+    )
+    if ensure_bos_token and tokenizer.bos_token_id is not None:
+        bos_token_id = tokenizer.bos_token_id
+    else:
+        bos_token_id = None
```

Comment:
Tiny nitpick -- you don't need to check the `bos_token_id` against `None` in the `if` statement if you're assigning to `None` in the else clause

Reply:
Do you think we should throw an error in the case where ensure_bos_token is True and tokenizer.bos_token_id is None?

Reply:
Mmm good question. To my knowledge, we're never calling this with ensure_bos_token set to False... In other words, I'm not sure the kwarg is really providing any value. I don't think an exception is really necessary in this case. What do you think?

Reply:
I'm just going to log a warning in that case for now